arch: arc: use the way of GEN_PRIV_STACK for privilege stack

vonhust opened the pull request:
drop the original C macro based allocation of privilged stack as
it may cause the waste of memory for ARC MPUv2.

now use the way of GEN_PRIV_STACK to generate privilege stack as
other archs did, e.g. ARM.


Fixes #17755

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: use the way of GEN_PRIV_STACK for privilege stack

drop the original C macro based allocation of privilged stack as
it may cause the waste of memory for ARC MPUv2.

now use the way of GEN_PRIV_STACK to generate privilege stack as
other archs did, e.g. ARM.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

andrewboie commented on arch/arc/core/mpu/Kconfig:
```diff
@@ -28,6 +28,7 @@ config ARC_MPU
 	select ARC_CORE_MPU
 	select THREAD_STACK_INFO
 	select MEMORY_PROTECTION
+	select GEN_PRIV_STACKS
```
shouldn't we only be doing this if ARC_MPU_VER = 2?
Not seeing benefit for later MPU revisions which don't have power-of-two alignment constraints.

andrewboie replied:
Note: ARM doesn't do this correctly either. There is an open issue #24048

vonhust replied:
I know GEN_PRIV_STACKS is not necessary for MPU_VER 3.  Enable it for V3 is to have the same logic for MPU drivers to simplify the code.  @ruuddw  @abrodkin , what's your opinion. If required ,  I can keep the original logic for MPU_VER 3

vonhust replied:
I add a commit to disable GEN_PRIV_STACK for ARC MPU v3

Commit:
arch: arc: ARC MPUv3 doesn't use GEN_PRIV_STACK

Because ARC MPUv3 doesn't have a strong alignment requirement
as ARC MPUv2 does, no use of GEN_PRIV_STACK for it.

Without GEN_PRIV_STACK, all stack elements can be in one stack object.
See #24048.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

andrewboie commented on include/arch/arc/arch.h:
```diff
@@ -87,55 +87,66 @@ extern "C" {
+#if  defined(CONFIG_USERSPACE)
+
+#if CONFIG_ARC_MPU_VER == 2
+/* MPU stack guard does not works for MPUv2 as it uses GEN_PRIV_STACK */
```
Strictly speaking, CONFIG_GEN_PRIV_STACKS could be made to reserve the needed space for the MPU guard in the privilege area. But the 2K minimum region size makes this, on a practical level, fairly useless. I appreciate simply removing support for MPU stack guards for V2, I never felt it was worth maintaining.

ruuddw replied:
Agree with Andrew, this comment is a bit confusing. Simply remove or state that MPU stack guards are not supported for V2 due to size overhead.

vonhust replied:
Ok, I will improve the comments. I agree with andrews, no need to have mpu stack guards even no userspace. Becasue it will consume 2K bytes.  ARC STACK CHECKING is a better option.

tyrantlizard77 reviewed (approved):
thanks @vonhust !

tyrantlizard77 reviewed (changes requested):
@vonhust how did you test this PR?
My local sanitycheck runs produce many failures with the `nsim_em_stack_guard` target.
Since this PR removes stack guard support for ARC MPU v2, this is not surprising. But you need to finish the removal by removing that build target.

tyrantlizard77 reviewed (changes requested):
Summary of changes that need to be made to this PR, all related to dropping support for MPU stack guards on ARC MPU v2:

1. arch/arc/core/mpu/Kconfig needs to be modified such that MPU_STACK_GUARD needs to be gated on ARC_MPU_VER != 2
2. nsim_em_mpu_stack_guard removed from Zephyr
3. The clause in include/arch/arc/arch.h line 71 which sets sets STACK_GUARD_SIZE 2048
4. Comments in include/arch/arc/arch.h like 136 which describes 2K stack guards with MPUv2

vonhust commented:
> Summary of changes that need to be made to this PR, all related to dropping support for MPU stack guards on ARC MPU v2:
> 
> 1. arch/arc/core/mpu/Kconfig needs to be modified such that MPU_STACK_GUARD needs to be gated on ARC_MPU_VER != 2
> 2. nsim_em_mpu_stack_guard removed from Zephyr
> 3. The clause in include/arch/arc/arch.h line 71 which sets sets STACK_GUARD_SIZE 2048
> 4. Comments in include/arch/arc/arch.h like 136 which describes 2K stack guards with MPUv2

Thanks, i will modify them.

Commit:
arch: arc: remove MPU_STACK_GUARD for ARC_MPU_VER 2

ARC_MPU_VER 2 has a strong requirement in
  * size, must be >= 2048 bytes and power of 2
  * start address must be aligned to size

It may bring a big waste of memory.

On the other hand, GEN_PRIV_STACK is used for ARC_MPU_VER 2,
it conflicts with MPU_STACK_GUARD.

So considering the limmitations, remove MPU_STACK_GUARD for
ARC_MPU_VER 2

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

andrewboie commented:
@vonhust do you think you will have time to update this in the next few days? some work I have pending is based on this...if not I can try to pick up where you left off

vonhust commented:
> @vonhust do you think you will have time to update this in the next few days? some work I have pending is based on this...if not I can try to pick up where you left off

@andrewboie  already updated

andrewboie commented:
@abrodkin @ruuddw I'd like to get this merged soon, as some work I am doing depends on this.

Meanwhile, I will incorporate these changes into my PR #24714

andrewboie commented:
Local testing met with success, didn't see any failures that weren't accounted for by known issues.

ruuddw commented on arch/arc/core/mpu/arc_mpu_v2_internal.h:
```diff
@@ -204,33 +204,14 @@ void arc_core_mpu_configure_thread(struct k_thread *thread)
+	 *   for stack guard
+	 * - better to use hardware stack checking as mpu stack guard will
+	     consume MPU entry.
+	 */
```
This #if now only has some comments inside -> suggest to change the ifdef and remove the else.
The comment could be moved/merged with the Kconfig option or so, as explanation why it's not supported.

vonhust replied:
OK

ruuddw commented on arch/arc/core/userspace.S:
```diff
@@ -98,8 +98,10 @@ SECTION_FUNC(TEXT, z_arc_userspace_enter)
-	/* start of privilege stack */
-	add blink, r5, CONFIG_PRIVILEGED_STACK_SIZE + STACK_GUARD_SIZE
+	/* get start of privilege stack */
+	ld blink, [r6, _thread_offset_to_priv_stack_start]
```
Some explanation/assumptions on the parameters passed through registers for this function would be good (r5 is mentioned above, but r6 not).

vonhust replied:
Ok

ruuddw commented on include/arch/arc/arch.h:
```diff
@@ -70,7 +70,7 @@ extern "C" {
 #if defined(CONFIG_MPU_STACK_GUARD)
 	#if CONFIG_ARC_MPU_VER == 2
-	#define STACK_GUARD_SIZE 2048
+	#define STACK_GUARD_SIZE 0
```
This combination isn't supported, looks like these complex nested #ifdef's can be simplified.

vonhust replied:
already done

ruuddw commented on include/arch/arc/arch.h:
```diff
@@ -89,55 +89,66 @@ extern "C" {
+ * - region size must be 32 bytes aligned
+ * - region start must be 32 bytes aligned
+ */
+#define Z_ARC_MPU_SIZE_ALIGN(size) STACK_SIZE_ALIGN(size)
```
For maintainability, please explain that STACK_SIZE_ALIGN takes care of the 32 byte alignement - or, simply put the math that can be checked/understood here directly.

andrewboie replied:
side note: I've rewritten and standardized this across in my in-progress PR #24714. We have new macros: 

ARCH_STACK_PTR_ALIGN - Required alignment of the stack pointer
ARCH_THREAD_STACK_OBJ_ALIGN(size) - Required alignment of the base address of the stack object
ARCH_THREAD_STACK_SIZE_ADJUST(size) - Round up a requested stack size to meet constraints (MPU granularity, page boundary, reserved data, etc.)

STACK_ALIGN / STACK_SIZE_ALIGN are completely removed. They were improperly namespaced and didn't have the same semantics everywhere.

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -153,7 +153,7 @@ config ARC_STACK_PROTECTION
 	default y if HW_STACK_PROTECTION
 	select ARC_STACK_CHECKING if ARC_HAS_STACK_CHECKING
-	select MPU_STACK_GUARD if (!ARC_STACK_CHECKING && ARC_MPU)
+	select MPU_STACK_GUARD if (!ARC_STACK_CHECKING && ARC_MPU && ARC_MPU_VER !=2)
```
Is MPU stack guard still supported for non Userspace builds?
Not really clear/consistent in the code:  Kconfig implies not, include/arc/arc/arch.h implies it is?

andrewboie replied:
In the general case: MPU/MMU stack guards and user mode are orthogonal features. 

You can have one or the other, both, or neither, there should be no dependencies.

ruuddw commented on include/arch/arc/arch.h:
```diff
@@ -89,55 +89,66 @@ extern "C" {
  * For MPU STACK_GUARD  kernel stacks do not need a MPU region to protect,
  * only guard needs to be protected and aligned. For MPUv3, MPU_STACK_GUARD
  * requires start 32 bytes aligned, also for size which is decided by stack
  * array and USERSPACE; For MPUv2, MPU_STACK_GUARD requires
```
Is MPU stack guard for no-userspace supported with MPU v2?

andrewboie replied:
stack guards have always been orthogonal to userspace, so I expect the answer is no.

vonhust replied:
no sense to have MPU stack guard for MPUv2, even no-userspace. Indeed, stack guards have always been orthogonal to userspace, but because of the hardware limitation of ARC MPUv2 (memory waste),  better to drop it, we can use ARC STACK CHECKING instead.

andrewboie replied:
@vonhust agreed, 2K of dead memory per thread is not a practical configuration.

vonhust commented:
ping for merge

andrewboie commented:
> ping for merge

@vonust I'd like some more +1s from Synopsys before doing so, @abrodkin @ruuddw can you re-visit this PR?

abrodkin commented:
I will approve this change but frankly I haven't spent enough time looking into technical details as I'm not up to speed yet with peculiarities of both HW & SW parts discussed here, so it's aproval from @ruuddw which will make real sense.

ruuddw reviewed (approved):
Ok for now. Some MPU/stack align macro logic can be further improved once PR #24714 has been completed.
